### PR TITLE
[PW-5093] Add handling for `cueUpdate` events

### PR DESCRIPTION
**Pull request by hanneslinder**
_Waiting until the player PR is finished before we merge this one_ 

This PR adds support for `cueUpdate` events

Changes to the UI: 

- Whenever `cueUpdate` event is triggered, we check if there is a previous cue in the `subtitleManager` that matches the id. 
- If a match is found we create a new subtitle label and replace the current one. 
- Matching by id works because `cueUpdate` events contain the same `start` and `text` properties as the `cueEnter` event, only the `html` content is different.

### CODE REVIEW

File: src/ts/components/subtitleoverlay.ts [hunk 1]
```diff
@@ -256,18 +280,20 @@ export class SubtitleOverlay extends Container<ContainerConfig> {
...
       this.updateComponents();
       this.show();
     }
   }
 
   removePreviewSubtitleLabel(): void {
-    this.previewSubtitleActive = false;
-    this.subtitleContainerManager.removeLabel(this.previewSubtitle);
-    this.updateComponents();
+    if (this.previewSubtitleActive) {
```
Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:292]: nit: That way you are not the owner of all lines, but only the updated IF clause
```suggestion
    if (!this.previewSubtitleActive) {
        return;
    }
```

File: src/ts/components/subtitleoverlay.ts [hunk 2]
```diff
@@ -339,34 +365,52 @@ class ActiveSubtitleManager {
...
     // Create array for id if it does not exist
     this.activeSubtitleCueMap[id] = this.activeSubtitleCueMap[id] || [];
 
     // Add cue
     this.activeSubtitleCueMap[id].push({ event, label });
     this.activeSubtitleCueCount++;
+  }
 
-    return label;
+  private removeCueFromMap(event: SubtitleCueEvent): SubtitleLabel {
```
Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:391]: If we apply the FIFO approach, we could then name it `popLastCueFromMap`

As remove suggest we are fully removing and it should not exist anymore.

Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:391]: As per code talk, we should use `undefined` instead of nulls
+ Missing return type for undefined
```suggestion
  private removeCueFromMap(event: SubtitleCueEvent): SubtitleLabel | undefined {
```

File: src/ts/components/subtitleoverlay.ts [hunk 3]
```diff
@@ -339,34 +365,52 @@ class ActiveSubtitleManager {
...
+       * works fine. Theoretically it can happen that two cues with colliding ids are removed at different times, in
+       * the wrong order. This rare case has yet to be observed. If it ever gets an issue, we can take the unstable
+       * cue end time (which can change between CueEnter and CueExit IN CueUpdate) and use it as an
+       * additional hint to try and remove the correct one of the colliding cues.
+       */
+      let activeSubtitleCue = activeSubtitleCues.shift();
+      this.activeSubtitleCueCount--;
+
+      return activeSubtitleCue.label;
+    } else {
```
Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:408]: Nit: No need for `else`, as the previous case has a return

File: src/ts/components/subtitleoverlay.ts [hunk 4]
```diff
@@ -339,34 +365,52 @@ class ActiveSubtitleManager {
...
+      return null;
```
Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:409]: As per code talk we should prefer undefined over nulls
```suggestion
      return undefined;
```

File: src/ts/components/subtitleoverlay.ts [hunk 5]
```diff
@@ -457,22 +483,27 @@ export class SubtitleRegionContainerManager {
...
+      return {
+        regionContainerId: label.vtt.region && label.vtt.region.id ? label.vtt.region.id : 'vtt',
+        regionName: 'vtt',
+      };
+    } else {
+      return {
+        regionContainerId: label.region || 'default',
+        regionName: label.region || 'default',
+      };
+    }
```
Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:497]: No need for ELSE as we have a return in the previous statement
```suggestion
    }
      return {
        regionContainerId: label.region || 'default',
        regionName: label.region || 'default',
      };
   
```

File: src/ts/components/subtitleoverlay.ts [hunk 6]
```diff
@@ -507,6 +538,13 @@ export class SubtitleRegionContainerManager {
     this.subtitleRegionContainers[regionContainerId].addLabel(label, overlaySize);
   }
 
+  updateLabel(previousLabel: SubtitleLabel, newLabel: SubtitleLabel): void {
+    const { regionContainerId } = this.getRegion(previousLabel);
+
+    this.subtitleRegionContainers[regionContainerId].removeLabel(previousLabel);
```
Comment (NikolaNikushev) [inline src/ts/components/subtitleoverlay.ts:544]: Question: here we implement a different logic than the FIFO approach above, is this correct?

**Review by NikolaNikushev — CHANGES_REQUESTED** [review]

File: src/ts/components/subtitleoverlay.ts [hunk 7]
```diff
@@ -54,27 +54,29 @@ export class SubtitleOverlay extends Container<ContainerConfig> {
...
       this.updateComponents();
     });
 
+    player.on(player.exports.PlayerEvent.CueUpdate, (event: SubtitleCueEvent) => {
+      const label = this.generateLabel(event);
+      const previousLabel = subtitleManager.getPreviousCue(event);
+
+      if (previousLabel) {
+        subtitleManager.cueUpdate(event, label);
+        this.subtitleContainerManager.updateLabel(previousLabel, label);
```
[the remark this answers is not in the document]
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts:76]: Sorry I am not following, mind to explain what you mean?
What's happening here is that in order to update a cue we must first find the previous one. If one is found we remove it from the DOM, and insert a new one.

File: src/ts/components/subtitleoverlay.ts (hunk 2 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: Thanks for spotting, fixed in https://github.com/bitmovin/bitmovin-player-ui/pull/378/commits/6b9d35304a117fb79d93fe0fad303c7e013b989d

File: src/ts/components/subtitleoverlay.ts (hunk 5 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: I agree, fixed in https://github.com/bitmovin/bitmovin-player-ui/pull/378/commits/6b9d35304a117fb79d93fe0fad303c7e013b989d

File: src/ts/components/subtitleoverlay.ts (hunk 4 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: Fixed in 6b9d353

File: src/ts/components/subtitleoverlay.ts (hunk 3 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: Fixed in 6b9d353

File: src/ts/components/subtitleoverlay.ts [hunk 8]
```diff
@@ -339,34 +365,52 @@ class ActiveSubtitleManager {
...
-      text: event.html || ActiveSubtitleManager.generateImageTagText(event.image) || event.text,
-      vtt: event.vtt,
-      region: event.region,
-      regionStyle: event.regionStyle,
-    });
+  cueUpdate(event: SubtitleCueEvent, label: SubtitleLabel): void {
+    this.addCueToMap(event, label);
+  }
+
+  getPreviousCue(event: SubtitleCueEvent): SubtitleLabel {
```
[the remark this answers is not in the document]
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: As discussed offline I would rather keep it like that as removal of cues from the internal map is of no concern to the outside world. Having to manage the cue map from outside would only increase complexity and invite bugs without any real gain.

File: src/ts/components/subtitleoverlay.ts [hunk 9]
```diff
@@ -54,27 +54,29 @@ export class SubtitleOverlay extends Container<ContainerConfig> {
...
-      this.subtitleContainerManager.addLabel(labelToAdd, this.getDomElement().size());
+      this.subtitleContainerManager.addLabel(label, this.getDomElement().size());
       this.updateComponents();
     });
 
+    player.on(player.exports.PlayerEvent.CueUpdate, (event: SubtitleCueEvent) => {
+      const label = this.generateLabel(event);
+      const previousLabel = subtitleManager.getPreviousCue(event);
+
+      if (previousLabel) {
```
[the remark this answers is not in the document]
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts:74]: I disagree, this should be kept as implementation detail inside the class.

File: src/ts/components/subtitleoverlay.ts (hunk 1 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: Changed in https://github.com/bitmovin/bitmovin-player-ui/pull/378/commits/144549952d0db2a91607a93623db00df517d7dcb

File: src/ts/components/subtitleoverlay.ts (hunk 2 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: Sure thing, I've renamed the method in https://github.com/bitmovin/bitmovin-player-ui/pull/378/commits/79883890938961fab7227b2b191bc2c8e77dcc38

File: src/ts/components/subtitleoverlay.ts (hunk 6 above)
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts:542]: Unless I am missing something the FIFO approach should still be valid here.

**Review by NikolaNikushev — APPROVED** [review]
LGTM

**Review by felix-hoc — COMMENTED** [review]
Nice and clean changes 💪🏼 
I'm just not 100% certain on the flow to remove the previous cue.

File: src/ts/components/subtitleoverlay.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts:74]: Yeah I agree that it's a bit hard to follow and I am open to suggestions on how to simplify this one :smiley: I didn't want to refactor the whole `ActiveSubtitleManager` so I tried to make it work with what was already there. If required I can still rework this part of the code tho, just let me know your thoughts.

[the remark this answers is not in the document]
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts]: Thanks @felix-hoc for your suggestion. I've applied it now and it indeed looks a bit cleaner that way :smiley:

**Review by felix-hoc — APPROVED** [review]
LGTM 👍🏼 Can confirm it works with the latest player build from the linked PR :)

**Review by stonko1994 — CHANGES_REQUESTED** [review]
- Please try to not put any internal links in the PR description as this repo is open source
- CEA captions seem to be broken.

You can test it using one of those buttons:
Screenshot 2021-05-04 at 08 55 27


# Expected behaviour (dev):
_I have a black background on the subtitles configured via the UI to make it visible_
## CEA-608 subtitle test pattern (60 secs)
Screenshot 2021-05-04 at 08 57 43

## CEA-608 subtitle test pattern 2
Screenshot 2021-05-04 at 08 57 59

# Current behaviour on this branch
## CEA-608 subtitle test pattern (60 secs)
- No output at all
## CEA-608 subtitle test pattern 2
- No output at all

File: src/ts/components/subtitleoverlay.ts [hunk 10]
```diff
@@ -339,34 +366,53 @@ class ActiveSubtitleManager {
...
+       */
+      let activeSubtitleCue = activeSubtitleCues.shift();
+      this.activeSubtitleCueCount--;
+
+      return activeSubtitleCue.label;
+    }
   }
 
-  private static generateImageTagText(imageData: string): string {
+  static generateImageTagText(imageData: string): string | undefined {
```
[the remark this answers is not in the document]
Comment (hanneslinder) [reply src/ts/components/subtitleoverlay.ts:415]: Never seen optional chaining used for a return type, is this valid?

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/) 
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
+## [develop]
+
+### Added
+- Support for subtitle `CueUpdate` events
```
[the remark this answers is not in the document]
Comment (hanneslinder) [reply CHANGELOG.md]: Updated the changelog to include the player version here https://github.com/bitmovin/bitmovin-player-ui/pull/378/commits/2b239ffe5253e59012ca5814329551e29d1171be

**Review by stonko1994 — COMMENTED** [review]
The CEA rendering looks wrong 🤔 

Screenshot 2021-05-04 at 10 41 43
Screenshot 2021-05-04 at 10 41 07

File: src/ts/components/subtitleoverlay.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (stonko1994) [reply src/ts/components/subtitleoverlay.ts:413]: https://github.com/microsoft/TypeScript/issues/13184

**Review by stonko1994 — APPROVED** [review]
Manual testing looks good now 👍

**Comment by jmsn** [comment]
I guess this should have gone to `develop` instead of `master` 😄
